dynamodb/src/main/java/org/apache/metamodel/dynamodb/DynamoDbDataContext.java

```diff
@@ -0,0 +1,320 @@
+import com.amazonaws.services.dynamodbv2.model.DescribeTableResult;
+import com.amazonaws.services.dynamodbv2.model.GetItemRequest;
+import com.amazonaws.services.dynamodbv2.model.GetItemResult;
+import com.amazonaws.services.dynamodbv2.model.GlobalSecondaryIndexDescription;
+import com.amazonaws.services.dynamodbv2.model.KeySchemaElement;
+import com.amazonaws.services.dynamodbv2.model.ListTablesResult;
+import com.amazonaws.services.dynamodbv2.model.LocalSecondaryIndexDescription;
+import com.amazonaws.services.dynamodbv2.model.ScanRequest;
+import com.amazonaws.services.dynamodbv2.model.ScanResult;
+import com.amazonaws.services.dynamodbv2.model.TableDescription;
+
+/**
+ * DataContext implementation for Amazon DynamoDB.
+ */
+public class DynamoDbDataContext extends QueryPostprocessDataContext implements UpdateableDataContext, Closeable {
+
+    /**
+     * System property key used for getting the read throughput capacity when
+     * creating new tables. Defaults to 5.
+     */
+    public static final String SYSTEM_PROPERTY_THROUGHPUT_READ_CAPACITY = "metamodel.dynamodb.throughput.capacity.read";
+
+    /**
+     * System property key used for getting the write throughput capacity when
+     * creating new tables. Defaults to 5.
+     */
+    public static final String SYSTEM_PROPERTY_THROUGHPUT_WRITE_CAPACITY = "metamodel.dynamodb.throughput.capacity.write";
+
+    private static final Logger logger = LoggerFactory.getLogger(DynamoDbDataContext.class);
+
+    /**
+     * The artificial schema name used by this DataContext.
+     */
+    public static final String SCHEMA_NAME = "public";
+
+    private final AmazonDynamoDB _dynamoDb;
+    private final boolean _shutdownOnClose;
+    private final SimpleTableDef[] _tableDefs;
+
+    public DynamoDbDataContext() {
+        this(AmazonDynamoDBClientBuilder.defaultClient(), null, true);
+    }
+
+    public DynamoDbDataContext(SimpleTableDef[] tableDefs) {
+        this(AmazonDynamoDBClientBuilder.defaultClient(), tableDefs, true);
+    }
+
+    public DynamoDbDataContext(AmazonDynamoDB client) {
+        this(client, null, false);
+    }
+
+    public DynamoDbDataContext(AmazonDynamoDB client, SimpleTableDef[] tableDefs) {
+        this(client, tableDefs, false);
+    }
+
+    private DynamoDbDataContext(AmazonDynamoDB client, SimpleTableDef[] tableDefs, boolean shutdownOnClose) {
+        _dynamoDb = client;
+        _tableDefs = (tableDefs == null ? new SimpleTableDef[0] : tableDefs);
+        _shutdownOnClose = shutdownOnClose;
+    }
+
+    public AmazonDynamoDB getDynamoDb() {
+        return _dynamoDb;
+    }
+
+    @Override
+    public void close() {
+        if (_shutdownOnClose) {
+            _dynamoDb.shutdown();
+        }
+    }
+
+    @Override
+    protected Schema getMainSchema() throws MetaModelException {
+        final Map<String, SimpleTableDef> tableDefs = new HashMap<>();
+        for (SimpleTableDef tableDef : _tableDefs) {
```

Comment:
I'll only mention it here, but all your enhanced for loops could have `final`s.

(as well as all or most of your method signatures, though I'm not sure I can convince you to do that :))